packages/firestore/src/local/indexeddb_schema.ts

```diff
@@ -21,54 +21,55 @@ import { ResourcePath } from '../model/path';
+import { PersistencePromise } from './persistence_promise';
+import { SnapshotVersion } from '../core/snapshot_version';
 
-export const SCHEMA_VERSION = 1;
-
-/** Performs database creation and (in the future) upgrades between versions. */
-export function createOrUpgradeDb(db: IDBDatabase, oldVersion: number): void {
-  assert(oldVersion === 0, 'Unexpected upgrade from version ' + oldVersion);
-
-  db.createObjectStore(DbMutationQueue.store, {
-    keyPath: DbMutationQueue.keyPath
-  });
-
-  // TODO(mikelehen): Get rid of "as any" if/when TypeScript fixes their
-  // types. https://github.com/Microsoft/TypeScript/issues/14322
-  db.createObjectStore(
-    DbMutationBatch.store,
-    // tslint:disable-next-line:no-any
-    { keyPath: DbMutationBatch.keyPath as any }
-  );
+/**
+ * Schema Version for the Web client (containing the Mutation Queue, the Query
+ * and the Remote Document Cache) and target count added to the target global.
+ */
+export const SCHEMA_VERSION = 2;
 
-  const targetDocumentsStore = db.createObjectStore(
-    DbTargetDocument.store,
-    // tslint:disable-next-line:no-any
-    { keyPath: DbTargetDocument.keyPath as any }
-  );
-  targetDocumentsStore.createIndex(
-    DbTargetDocument.documentTargetsIndex,
-    DbTargetDocument.documentTargetsKeyPath,
-    { unique: true }
+/**
+ * Performs database creation and schema upgrades.
+ *
+ * Note that in production, this method is only ever used to upgrade the schema
+ * to SCHEMA_VERSION. Different versions are only used for testing and
+ * local feature development.
+ */
+export function createOrUpgradeDb(
+  db: IDBDatabase,
+  txn: IDBTransaction,
+  fromVersion: number,
+  toVersion: number
+): PersistencePromise<void> {
+  // This function currently supports migrating to schema version 1 (Mutation
+  // Queue, Query and Remote Document Cache) and schema version 2 (Query
+  // counting).
+  assert(
+    fromVersion < toVersion && fromVersion >= 0 && toVersion <= 2,
+    'Unexpected schema upgrade from v${fromVersion} to v{toVersion}.'
   );
 
-  const targetStore = db.createObjectStore(DbTarget.store, {
-    keyPath: DbTarget.keyPath
-  });
-  // NOTE: This is unique only because the TargetId is the suffix.
-  targetStore.createIndex(
-    DbTarget.queryTargetsIndexName,
-    DbTarget.queryTargetsKeyPath,
-    { unique: true }
-  );
+  if (fromVersion < 1 && toVersion >= 1) {
+    createOwnerStore(db);
+    createMutationQueue(db);
+    createQueryCache(db);
+    createRemoteDocumentCache(db);
+  }
 
-  // NOTE: keys for these stores are specified explicitly rather than using a
-  // keyPath.
-  db.createObjectStore(DbDocumentMutation.store);
-  db.createObjectStore(DbRemoteDocument.store);
-  db.createObjectStore(DbOwner.store);
-  db.createObjectStore(DbTargetGlobal.store);
+  let p = PersistencePromise.resolve();
+  if (fromVersion < 2 && toVersion >= 2) {
+    p = p.next(() => ensureTargetGlobal(txn)).next(() => saveTargetCount(txn));
```

Comment:
Minor but I'd rather combine these two methods and have saveTargetCount just gracefully handle if the row is missing. This 1) Avoids unnecessarily doing a get-put-get-put, and 2) Makes it clearer what is actually new in schema version 2 (the target count).  If you want, you can keep ensureTargetGlobal() but make it return the TargetGlobal row, and then call it from within saveTargetCount().

Also, I suggest renaming saveTargetCount() to populateTargetCount() to make it clearer that it needs to do work to figure out what the target count needs to be.

Reply:
Renamed to `populateTargetCount`. I was initially going to keep the return value of `ensureTargetGlobal` as `void`, since they are independent steps (making sure the global exists could be its own schema migration), but since both you and @schmidt-sebastian asked for it and I don't feel too strongly about it, I'll make the change.

Reply:
I don't see the rename...  I also still don't think there's a need to have two steps here.  TargetGlobal was an existing part of the schema in v1. The only thing you added was a new field to it.

Another way to deal with this would be to get rid of ensureTargetGlobal() and have populateTargetCount() just do nothing if the TargetGlobal row doesn't exist yet (since it should get populated correctly as 0 when the TargetGlobal row gets created as a normal part of the client operating).  And actually, I think I like this better since 1) it avoids us duplicating the default `new DbTargetGlobal( ... )` code.